### ISSUE TITLE
mini_batch_sparse_pca not working

### ISSUE THREAD

**Issue by thomasaarholt** [hyperspy/hyperspy#1361]
`s.decomposition(True, algorithm="mini_batch_sparse_pca", output_dimension=5)` doesn't work. Thought I'd try it to see what it would return for my EDX map, but it returned an empty array of loadings and factors.
I think we should really update the documentation to reflect the new changes to PCA. Perhaps with the addition of ORNMF we could do a proper update of at least the `algorithm` argument descriptor?

**Comment by thomasaarholt** [comment]
~~I may have screwed up, working on the wrong branch. I'll reopen this if it still isn't working later.~~
It's real.

**Comment by tjof2** [comment]
> but it returned an empty array of loadings and factors.

I think this is because you are using the default kwargs for [MiniBatchSparsePCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.MiniBatchSparsePCA.html). The same applies to the `'sparse_pca'` algorithm option too.

In particular, the `alpha` parameter. Try setting it very small to 1e-5 or so. For example, using the default values

```
>>> s.decomposition(algorithm="mini_batch_sparse_pca", output_dimension=5)
>>> loadings = s.get_decomposition_loadings().data
>>> factors = s.get_decomposition_factors().data
>>> print(loadings.mean())
>>> print(factors.mean())
>>> print(np.mean(factors == 0))

0.0
0.0
1.0 (i.e. all values are 0)
```

while

```
>>> s.decomposition(algorithm="mini_batch_sparse_pca", output_dimension=5, alpha=1e-5)
>>> loadings = s.get_decomposition_loadings().data
>>> factors = s.get_decomposition_factors().data
>>> print(loadings.mean())
>>> print(factors.mean())
>>> print(np.mean(factors == 0))

0.007826386225173738
0.18074411912790012
0.2896505376344086
```

> *alpha* : Sparsity controlling parameter. Higher values lead to sparser components.

A value of 1 is quite high, hence all the zeros

**Comment by tjof2** [comment]
> *alpha* : Sparsity controlling parameter. Higher values lead to sparser components.
A value of 1 is quite high, hence all the zeros

**Comment by tjof2** [comment]
Heuristically, we could set `alpha` to `1/sqrt(n_features)` if it  is not given explicitly, and pass through to sklearn.

**Comment by tjof2** [comment]
Certainly from a documentation point-of-view, this is something to be addressed in #1159

**Comment by tjof2** [comment]
@thomasaarholt is it fair to say this should be closed in favour of #1159? Since the algorithm does actually work as intended, but is poorly documented.

**Comment by thomasaarholt** [comment]
Yep, agreed.
